workmail-save-and-update-email/src/app.py

```diff
@@ -0,0 +1,122 @@
+        }
+
+    context: object, required
+    Lambda Context runtime methods and attributes. See https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+
+    Returns
+    -------
+    Amazon WorkMail Sync Lambda Response Format
+    For more information, see https://docs.aws.amazon.com/workmail/latest/adminguide/lambda.html#synchronous-schema
+        return {
+          'actions': [                                              # Required, should contain at least 1 list element
+          {
+            'action' : {                                            # Required
+              'type': 'string',                                     # Required. Can be "BOUNCE", "DROP" or "DEFAULT"
+              'parameters': { <various> }                           # Optional. For bounce, <various> can be {"bounceMessage": "message that goes in bounce mail"}
+            },
+            'recipients': list of strings,                          # Optional. Indicates list of recipients for which this action applies
+            'default': boolean                                      # Optional. Indicates whether this action applies to all recipients
+          }
+        ]}
+
+    """
+
+    logger.info(f"Received event: {event}")
+    email_from = event['envelope']['mailFrom']
+    recipients = event['envelope']['recipients']
+    message_id = event['messageId']
+    key = str(uuid.uuid4())
+
+    # Determine if the message is internal
+    if utils.extract_domains([email_from]) == utils.extract_domains(recipients):
+        internal_message = True
+    else:
+        internal_message = False
+        
+    update_internal_msg = (utils.get_env_var('UPDATE_INTERNAL_MESSAGES') == 'True')
+    update_external_msg = (utils.get_env_var('UPDATE_EXTERNAL_MESSAGES') == 'True')
+    save_and_update_msg = False
+
+    if internal_message:
+        if update_internal_msg:
+            save_and_update_msg = True
+    else:
+        if update_external_msg:
+            save_and_update_msg = True
+
+    try:
+        # 1. Download email
+        downloaded_email = utils.download_email(message_id)
+        # 2. Save and update original email
+        if save_and_update_msg:
+            # 3. Save the orginal, unmodified, email message source
+            utils.save_email(downloaded_email, key)
+            # 4. Save the event data (metadata) about the message so we know the envelope details that aren't in the message source
+            utils.save_email_metadata(event, key)
+            # 5. Update the email with the desired modifications
+            updated_email = utils.update_email(downloaded_email, event['subject'], event['flowDirection'], key)
+            logger.info("Providing modified message for WorkMail")
+            utils.update_workmail(message_id, updated_email, key)
+        else:
+            logger.info("Preserving original message for WorkMail")
+            utils.update_workmail(message_id, downloaded_email, key)
```

Comment:
I'm not sure if I understand the intent here. If the message is not modified WorkMail will automatically use the original message for delivery

Reply:
I think I just did not understand how to return back to workmail synchronous unmodified and without causing the messages to timeout and treated as spam. I will check the docs.